algorithms/src/snark/marlin/marlin.rs

```diff
@@ -75,85 +77,107 @@ impl<E: PairingEngine, FS: AlgebraicSponge<E::Fq, 2>, MM: MarlinMode> MarlinSNAR
+                circuit_commitments,
+                verifier_key,
+                mode: PhantomData,
+                hash: indexed_circuit.hash.clone(),
+            };
+            circuit_verifying_keys.push(circuit_verifying_key.clone());
+
+            let circuit_proving_key = CircuitProvingKey {
+                circuit: Arc::new(indexed_circuit),
+                circuit_commitment_randomness,
+                circuit_verifying_key: circuit_verifying_key,
+                committer_key: Arc::new(committer_key),
+            };
+            circuit_proving_keys.push(circuit_proving_key);
+        }
 
-        let commit_time = start_timer!(|| "Commit to index polynomials");
-        let (mut circuit_commitments, circuit_commitment_randomness): (_, _) =
-            SonicKZG10::<E, FS>::commit(&committer_key, index.iter().map(Into::into), None)?;
-        end_timer!(commit_time);
-
-        circuit_commitments.sort_by(|c1, c2| c1.label().cmp(c2.label()));
-        let circuit_commitments = circuit_commitments.into_iter().map(|c| *c.commitment()).collect();
-        let circuit_verifying_key = CircuitVerifyingKey {
-            circuit_info: index.index_info,
-            circuit_commitments,
-            verifier_key,
-            mode: PhantomData,
-        };
+        end_timer!(index_time);
 
-        let circuit_proving_key = CircuitProvingKey {
-            circuit: Arc::new(index),
-            circuit_commitment_randomness,
-            circuit_verifying_key: circuit_verifying_key.clone(),
-            committer_key: Arc::new(committer_key),
-        };
+        Ok((circuit_proving_keys, circuit_verifying_keys))
 
-        end_timer!(index_time);
+    }
 
-        Ok((circuit_proving_key, circuit_verifying_key))
+    /// Generates the circuit proving and verifying keys.
+    /// This is a deterministic algorithm that anyone can rerun.
+    #[allow(clippy::type_complexity)]
+    pub fn circuit_setup<C: ConstraintSynthesizer<E::Fr>>(
+        universal_srs: &UniversalSRS<E>,
+        circuit: &C,
+    ) -> Result<(CircuitProvingKey<E, MM>, CircuitVerifyingKey<E, MM>), SNARKError> {
+        let (mut proving_keys, mut verifying_keys) = Self::batch_circuit_setup(universal_srs, &[circuit])?;
+        assert!(proving_keys.len() == 1 && verifying_keys.len() == 1);
+        Ok((proving_keys.remove(0), verifying_keys.remove(0)))
     }
 
     fn terminate(terminator: &AtomicBool) -> Result<(), MarlinError> {
         if terminator.load(Ordering::Relaxed) { Err(MarlinError::Terminated) } else { Ok(()) }
     }
 
-    fn init_sponge(
+    fn init_sponge<'a>(
         fs_parameters: &FS::Parameters,
-        batch_size: usize,
-        circuit_commitments: &[crate::polycommit::sonic_pc::Commitment<E>],
-        inputs: &[Vec<E::Fr>],
+        batch_sizes: &BTreeMap<CircuitId, usize>,
+        circuit_commitments: &[Vec<crate::polycommit::sonic_pc::Commitment<E>>],
+        inputs: &BTreeMap<&CircuitId, &[Vec<E::Fr>]>,
```

Comment:
Let's wrap this into one map, i.e. `BTreeMap<CircuitId, (usize, Vec<Commitment>, &[Vec<E::Fr>])>`